packages/Discretization/src/DTK_DetailsNearestNeighborOperatorImpl.hpp

```diff
@@ -0,0 +1,148 @@
+        Kokkos::parallel_for(
+            REGION_NAME( "setup_queries" ),
+            Kokkos::RangePolicy<ExecutionSpace>( 0, n_target_points ),
+            KOKKOS_LAMBDA( int i ) {
+                nearest_queries( i ) = Details::nearest(
+                    Point{{target_points( i, 0 ), target_points( i, 1 ),
+                           target_points( i, 2 )}} );
+            } );
+        Kokkos::fence();
+        return nearest_queries;
+    }
+
+    static void
+    pullSourceValues( Teuchos::RCP<const Teuchos::Comm<int>> const &comm,
+                      Kokkos::View<double *, DeviceType> source_values,
+                      Kokkos::View<int *, DeviceType> &buffer_indices,
+                      Kokkos::View<int *, DeviceType> &buffer_ranks,
+                      Kokkos::View<double *, DeviceType> &buffer_values )
+    {
+        int const n_exports = buffer_indices.extent( 0 );
+        Tpetra::Distributor distributor( comm );
+        int const n_imports =
+            distributor.createFromSends( Teuchos::ArrayView<int>(
+                buffer_ranks.data(), buffer_ranks.extent( 0 ) ) );
+
+        Kokkos::View<int *, DeviceType> export_target_indices( "target_indices",
+                                                               n_exports );
+        iota( export_target_indices );
+        Kokkos::View<int *, DeviceType> import_target_indices( "target_indices",
+                                                               n_imports );
+        DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
+            distributor, export_target_indices, import_target_indices );
+
+        Kokkos::View<int *, DeviceType> export_source_indices = buffer_indices;
+        Kokkos::View<int *, DeviceType> import_source_indices( "source_indices",
+                                                               n_imports );
+        DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
+            distributor, export_source_indices, import_source_indices );
+
+        Kokkos::View<int *, DeviceType> export_ranks( "ranks", n_exports );
+        Kokkos::View<int *, DeviceType> import_ranks( "ranks", n_imports );
+        int const comm_rank = comm->getRank();
+        Kokkos::deep_copy( export_ranks, comm_rank );
+        DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
+            distributor, export_ranks, import_ranks );
+
+        buffer_indices = import_target_indices;
+        buffer_ranks = import_ranks;
+        Kokkos::realloc( buffer_values, n_imports );
+        Kokkos::parallel_for(
+            REGION_NAME( "get_source_values" ),
+            Kokkos::RangePolicy<ExecutionSpace>( 0, n_imports ),
+            KOKKOS_LAMBDA( int i ) {
+                buffer_values( i ) =
+                    source_values( import_source_indices( i ) );
+            } );
+        Kokkos::fence();
+    }
+
+    static void
+    pushTargetValues( Teuchos::RCP<const Teuchos::Comm<int>> const &comm,
```

Comment:
This functions does not respect `(in, out)`. Also the `const` are missing (you are not modifying `buffer_indices`,`buffer_ranks`, and `buffer_values`.

Reply:
Actually I am changing them so it cannot be declared `const`.

Reply:
Where? You change them in `pull` but not in `push`

Reply:
You are right.

Reply:
Done

Reply:
Still does not respect the coding conventions `target_values` should be the latest argument,

Reply:
Fixed (for real this time)